### PR TITLE
Update readme and docs.

### PR DESCRIPTION
**Pull request by sandreim**
Signed-off-by: Andrei Sandu <sandreim@amazon.com>

## Reason for This PR

Add missing documentation.

## Description of Changes

Add missing documentation.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,70 @@
...
+        .new_version()
+        .set_type_version(Test::type_id(), 3);
+
+    let test_struct = Test {
+        a: 1337,
+        b: 0xFF,
+        c: "c_value".to_owned(),
+    };
+
+    // Serialize to version 2- field c will not be serialized.
```
Comment (acatangiu) [inline README.md]: nit: 
```suggestion
    // Serialize to version 2 - field c will not be serialized.
```

File: src/lib.rs [hunk 2]
```diff
@@ -2,10 +2,22 @@
...
 #![deny(missing_docs)]
 
-//! Exports the Versionize derive proc macro that generates the Versionize implementation
-//! for structs, and enums by using annotations.
-
-extern crate proc_macro;
+//! Exports the `Versionize` derive proc macro that generates the Versionize trait 
+//! implementation.
+//! 
+//! Versionize generates serialization and deserialization code only for structures and unions.
```
Comment (acatangiu) [inline src/lib.rs]:
```suggestion
//! Versionize generates serialization and deserialization code only for structures and enums.
```

File: src/lib.rs [hunk 3]
```diff
@@ -30,7 +42,143 @@ pub(crate) const SEMANTIC_DE_FN: &str = "de_fn";
 pub(crate) const START_VERSION: &str = "start";
 pub(crate) const END_VERSION: &str = "end";
 
-/// Implements the derive proc macro.
+/// Generates serialization an deserialization code as an implementation of
```
Comment (acatangiu) [inline src/lib.rs]:
```suggestion
/// Generates serialization and deserialization code as an implementation of
```

File: src/lib.rs [hunk 4]
```diff
@@ -30,7 +42,143 @@ pub(crate) const SEMANTIC_DE_FN: &str = "de_fn";
...
+/// }
+/// 
+/// impl TestStruct {
+///     fn default_b(_source_version: u16) -> u8 {
+///         12u8
+///     }
+/// }
+/// ```
+/// 
+/// The function name needs to be specified as a string and it's prototype must
```
Comment (acatangiu) [inline src/lib.rs]:
```suggestion
/// The function name needs to be specified as a string and its prototype must
```

File: src/lib.rs [hunk 5]
```diff
@@ -30,7 +42,143 @@ pub(crate) const SEMANTIC_DE_FN: &str = "de_fn";
...
+/// take an u16 source version parameter and return a value of the same type as
+/// as the field.
+/// 
+/// #### start/end
+/// 
+/// Defines the field version lifetime. Fields can be added by specifing the start
+/// version of the structure when first defining them and can be later on removed 
+/// from serialization logic by adding and end version.
```
Comment (acatangiu) [inline src/lib.rs]:
```suggestion
/// from serialization logic by adding an end version. 
```

File: src/lib.rs [hunk 6]
```diff
@@ -30,7 +42,143 @@ pub(crate) const SEMANTIC_DE_FN: &str = "de_fn";
...
+/// 
+/// #### ser_fn
+/// * Not supported for enums. *
+/// 
+/// Defines a semantic serialization function for a field. The function needs
+/// to be specified as a string and implemented as a method attached to 
+/// the structure. The prototype of the function is `fn(&mut self, u16) -> VersionizeResult<()>`. 
+/// 
+/// If defined, the method is called when the field is skipped from serialization
+/// because it does not exist in the target version of the structure. It's implementation
```
Comment (acatangiu) [inline src/lib.rs]:
```suggestion
/// because it does not exist in the target version of the structure. Its implementation
```

File: src/lib.rs [hunk 7]
```diff
@@ -30,7 +42,143 @@ pub(crate) const SEMANTIC_DE_FN: &str = "de_fn";
...
+/// #### de_fn
+/// * Not supported for enums. *
+/// 
+/// Defines a semantic deserialization function for a field. The function needs
+/// to be specified as a string and implemented as a method attached to 
+/// the structure. The prototype of the function is `fn(&mut self, u16) -> VersionizeResult<()>`. 
+/// 
+/// If defined, the method is called if the field is skipped from deserialization
+/// because it does not exist in the source version of the serialized structure. 
+/// It's implementation can perform any mutation of `self` or return an error to stop
```
Comment (acatangiu) [inline src/lib.rs]:
```suggestion
/// Its implementation can perform any mutation of `self` or return an error to stop 
```

File: README.md [hunk 8]
```diff
@@ -1,9 +1,70 @@
-# versionize_derive
+**Versionize is a framework for version tolerant serializion/deserialization of Rust data structures, designed for usecases that need fast deserialization times and minimal size overhead. It does not aim to be a generic serialization framework and only the [bincode](https://crates.io/crates/bincode) backend is supported.**
```
Comment (georgepisaltu) [inline README.md]: Lines in markdown documents should not exceed a length of 80 characters. There are more lines that break this rule below as well.

File: src/lib.rs [hunk 9]
```diff
@@ -30,7 +42,143 @@ pub(crate) const SEMANTIC_DE_FN: &str = "de_fn";
...
+/// 
+/// Defines the field version lifetime. Fields can be added by specifing the start
+/// version of the structure when first defining them and can be later on removed 
+/// from serialization logic by adding and end version. 
+/// 
+/// For example: `#[version(start = 2, end = 4)]`. The field would be present in 
+/// the structure v2 and v3, but starting with v4 it would no longer be serialized
+/// or deserialized. 
+/// 
+/// Once a field is removed, it can never be added again at future version.
```
Comment (berciuliviu) [inline src/lib.rs]: Nit: added again **in a** future version

Comment (berciuliviu) [inline src/lib.rs]: Just a curiosity, if you want to add back a removed field, it should be added with a different name in the best case?

Comment (sandreim) [reply src/lib.rs]: Yes, that's one way of doing it.

File: README.md (hunk 8 above)
Comment (sandreim) [reply README.md]: fixed it.

**Comment by sandreim** [comment]
@acatangiu @georgepisaltu @berciuliviu Fixed all. PTAL.

**Review by acatangiu — DISMISSED** [review]
LGTM

File: README.md [hunk 10]
```diff
@@ -1,9 +1,81 @@
...
+
+#### You may be looking for:
+- [Versionize documentation](https://docs.rs/versionize/)
+- [Releases](https://github.com/firecracker-microvm/versionize_derive/releases)
 
 ## Important note
 
-This crate is currently used for cross-version serialization with the [Firecracker snapshot-restore dev preview](https://github.com/firecracker-microvm/firecracker/tree/v0.23.0), but has not been tested for other use cases. It should be considered **experimental software** outside the Firecracker context. It’s likely that this crate will see both interface and implementation changes in the future.
+This crate is currently used for cross-version serialization with the 
+[Firecracker snapshot-restore dev preview](https://github.com/firecracker-microvm/firecracker/tree/v0.24.0),
```
Comment (georgepisaltu) [inline README.md]: This line still breaks the rule. Here is a [quick guide](https://www.rubycoloredglasses.com/2017/11/markdown-links-80-character-line-length/) on how to fix it.

**Comment by sandreim** [comment]
@acatangiu @georgepisaltu @berciuliviu PTAL.

**Review by berciuliviu — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]

**Pull request merged**